### ISSUE TITLE
[Feature Request] Auto update plugins under $ZSH_CUSTOM folder

### ISSUE THREAD

**Issue by SukkaW** [ohmyzsh/ohmyzsh#9512]
**Is your feature request related to a problem? Please describe.**

oh-my-zsh built-in plugins will be updated during `omz update` (and auto-update as well) since those plugins are hosted in the `ohmyzsh` repo. However, there might be many users' plugins that are located under the `$ZSH_CUSTOM` folders. oh-my-zsh should update them as well when `omz update`.

**Describe the solution you'd like**

First, there should be an option (maybe `$UPDATE_ZSH_CUSTOM`). When enabled, `omz update` should look through every directory under the `$ZSH_CUSTOM` folder. If there is a `.git`, execute `git pull --rebase`.

**Describe alternatives you've considered**

There are many existed oh-my-zsh plugins that accomplish the same thing. But they heavily rely on overrides oh-my-zsh internal function. oh-my-zsh could provide an `after_update_hook` for them.

**Additional context**

- https://github.com/TamCore/autoupdate-oh-my-zsh-plugins
- https://github.com/tautomer/update-ohmyzsh-custom-plugins

Two oh-my-zsh plugins pop up in Google search result when I type in "oh my zsh custom plugin update".

**Comment by SukkaW** [comment]
Another alternative: `omz plugin update`.

**Comment by K-E-N-W-A-Y** [comment]
I used the below snippet from [this solution](https://unix.stackexchange.com/questions/477258/how-to-auto-update-custom-plugins-in-oh-my-zsh) before.

`# Custom Plugin update custom function
printf "\n${BLUE}%s${RESET}\n" "Updating custom plugins and themes"
cd custom/
for plugin in plugins/*/ themes/*/; do
  if [ -d "$plugin/.git" ]; then
     printf "${YELLOW}%s${RESET}\n" "${plugin%/}"
     git -C "$plugin" pull
  fi
done`

The problem with this hack is now I'm getting git merge errors. So I have to git stash & then update. So an official feature would be great.

**Comment by SukkaW** [comment]
> I used the below snippet from [this solution](https://unix.stackexchange.com/questions/477258/how-to-auto-update-custom-plugins-in-oh-my-zsh) before.
> The problem with this hack is now I'm getting git merge errors. So I have to git stash & then update. So an official feature would be great.

I don't need any "third-party" solution bacause I already got one (even better than the one your have found):

```bash
update_ohmyzsh_custom_plugins() {
    red=$(tput setaf 1)
    blue=$(tput setaf 4)
    green=$(tput setaf 2)
    reset=$(tput sgr0)

    echo ""
    printf "${blue}%s${reset}\n" "Upgrading custom plugins"

    find "${ZSH_CUSTOM:-~/.oh-my-zsh/custom}" -type d -name ".git" | while read LINE; do
        p=${LINE:h}
        pushd -q "${p}"

        if git pull --rebase; then
            printf "${green}%s${reset}\n" "${p:t} has been updated and/or is at the current version."
        else
            printf "${red}%s${reset}\n" "There was an error updating ${p:t}. Try again later?"
        fi
        popd -q
    done
}
```

The code snippet above will ignore any git merge errors since it uses `git pull --rebase`.

**Comment by K-E-N-W-A-Y** [comment]
@SukkaW I see that looks great. I mean official feature would be great. anyway, thank you for the snippet :)

**Comment by mr-salty** [comment]
yeah, my upgrade.sh is always in a modified state and I'd like that not to be the case.

I'd vote for just incorporating the above logic into `tools/upgrade.zsh` but we could also do what I'm doing; provide a hook and leave it up to the user:

```
local custom_upgrade=${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/upgrade.zsh
if [[ $ret -eq 0 && -x $custom_upgrade ]]; then
  $custom_upgrade
  ret=$?
fi
```

Previously I had hand-crafted `custom/upgrade.zsh` but I like @SukkaW 's version better so I'm using that now.  One note:  your expansion doesn't work if `$ZSH_CUSTOM` is unset - the `~` is interpreted literally and not expanded to `$HOME`; removing the double quotes fixes that although there may be a better fix.

**Comment by rea1shane** [comment]
Is this feature under development? I saw #10772 still draft.

**Comment by carlosala** [comment]
Yeah, it is on our mind, but it has been deprioritized at the moment. Let's see if we can push for it at some point.
